tuxedo-core/src/verifier/htlc.rs

```diff
@@ -0,0 +1,380 @@
+        }
+    }
+}
+
+impl Verifier for BlakeTwoHashLock {
+    type Redeemer = Vec<u8>;
+    fn verify(&self, _: &[u8], _: u32, secret: &Self::Redeemer) -> bool {
+        BlakeTwo256::hash(secret) == self.hash_lock
+    }
+}
+
+/// Allows a UTXO to be spent, and therefore acknowledged by an intended recipient by revealing
+/// a hash preimage. After an initial claim period elapses on chain, the UTXO can also be spent
+/// by the refunder. In practice, the refunder is often the same address initially funded the HTLC.
+///
+/// The receiver and refunder are specified as a simple public keys for simplicity. It would be
+/// interesting to use public key hash, or better yet, simply abstract this over some opaque
+/// inner verifier for maximum composability.
+///
+/// @lederstrumpf when the time expires, is the primary "happy" path supposed to remain open?
+/// Or is it that if the swap hasn't started on time, the refund is the only option.
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone, TypeInfo)]
+pub struct HashTimeLockContract {
+    /// The hash whose preimage must be revealed (along with the recipient's signature) to spend the UTXO.
+    pub hash_lock: H256,
+    /// The pubkey that is intended to receive and acknowledge receipt of the funds.
+    pub recipient_pubkey: Public,
+    /// The time (as a block height) when the refund path opens up.
+    pub claim_period_end: u32,
+    /// The address who can spend the coins without revealing the preimage after the claim period has ended.
+    pub refunder_pubkey: Public,
+}
+
+///
+#[derive(Serialize, Deserialize, Encode, Decode, Debug, PartialEq, Eq, Clone)]
+pub enum HtlcSpendPath {
+    ///
+    Claim {
+        secret: Vec<u8>,
+        signature: Signature,
+    },
+    ///
+    Refund { signature: Signature },
+}
+
+impl Verifier for HashTimeLockContract {
+    type Redeemer = HtlcSpendPath;
+
+    fn verify(&self, simplified_tx: &[u8], block_height: u32, spend_path: &HtlcSpendPath) -> bool {
+        match spend_path {
+            HtlcSpendPath::Claim { secret, signature } => {
+                // Claims are valid as long as the secret is correct and the receiver signature is correct.
+                BlakeTwo256::hash(secret) == self.hash_lock
+                    && sp_io::crypto::sr25519_verify(
+                        signature,
+                        simplified_tx,
+                        &self.recipient_pubkey,
+                    )
+            }
+            HtlcSpendPath::Refund { signature } => {
+                // Check that the time has elapsed
+                block_height >= self.claim_period_end
+
+                &&
+
+                // Check that the refunder has signed properly
+                sp_io::crypto::sr25519_verify(
+                    signature,
+                    simplified_tx,
+                    &self.refunder_pubkey,
+                )
+            }
+        }
+    }
+}
```

Comment:
What's your stance on composable `Verifier`?

I'm otherwise undecided whether you are deliberately not composing `HashLock` & `TimeLock` implementations to construct `HashTimeLockContract`, or are leaving this for a later stage.
I think there are pros (e.g. leaner interface & codebase) but also cons (more moving parts beneath a Verifier, so requires more careful versioning) to composing.
See https://github.com/Off-Narrative-Labs/Tuxedo/commit/29a20fe1 for a candidate implementation that reuses `Verifier` implementations from `HashLock` & `TimeLock`.

Reply:
Thanks for the detailed writeup and citations regarding expiring transactions. I'm not all that convinced by the Prophet's argument TBH. As others explained in the forum, there are already double spend scenarios that make a transaction invalid from one fork to another, so I don't see expiry being that much worse. The state boat from expired transactions is more convincing to me, but in the case of swaps, the transaction never truly expires, only one of its paths does, so I'm not too concerned. I think for now we let the happy path close for the reasons you explained. Anyone could write their own verifier that leaves it open or adds a `happy_path_expires: bool` field if they want it later.

I think composing the verifiers is a good idea. I was already starting to consider it when I was writing this PR. Even looking at the simple `HashLock`, for example, even it could be composed with an inner `Sr25519Signature` rather than repeating the sig check logic. And by making them composable, we can also make them generic. I could imagine some verifier combinators for scenarios like `And` `Or` `Not` etc. I'll experiment with it.

Reply:
I love the composed Verifier. Although it's not really more convenient in this particular case, it is actually a great example of what I believe should be a good practice in action.